### PR TITLE
Stage 2/6: Several directions

### PR DESCRIPTION
**Pull request by rabestro**
# Stage 2/6: Several directions
Project: [HyperMetro](https://hyperskill.org/projects/120)

## Description
It's time to expand our metro! To provide the citizens with more efficient transportation, we will extend our existing metro line and start a new one. Since now we will have several lines, there should be a possibility to choose which one to use. Also, let's not forget that trains go in both directions (except for the end stations). In this case, it makes sense to use a linked list with access to the previous element: a doubly-linked list. Its main feature, as you might have guessed, is that each element stores pointers to both the previous and the next element.

Since the file stores, not just one metro line but two, it is easier to organize the data by splitting it up between the two lines. In this case, we use a JSON file. It is a popular data-interchange format consisting of a map, where a key is a string type and a value can be any type. A key is separated from a value by a colon, and different key-value pairs are separated by a comma. You can find a comprehensive JSON tutorial at qhmit.com. By default, Java has no functionality to work with the JSON format, but Google made a special library GSON, which you can use in this project: see zetcode.com for more information on GSON.

## Objective
- Switch to using a JSON file instead of a simple text file.
- Add new stations in the given order
- The program should have the following commands: `/append "[line name]" "[station name]"`, `/add-head "[line name]" "[station name]"`, and `/remove "[line name]" "[station name]"`. Append adds a new station at the end and of the line, while add-head adds a new station at the beginning. Note that if the name of the line or station consists of several words, you should write it in quotation marks (after, they should be excluded from the name). If the name of a station consists of one word it should be parseable with or without quotes.
- The program should print a line that has been chosen with the command `/output "[line name]"`.
- The program should handle the incorrect format of the file records and output the message `Incorrect file`.
- Also, your program should handle the incorrect format of the command and print the message `Invalid command`.
If the path to the file is invalid, print an error message `Error! Such a file doesn't exist!`.
Add a special command `/exit` to finish the work of the program.

### CODE REVIEW

**Review by rabestro — COMMENTED** [review]
Решение второй стадии проекта HyperMetro.
https://hyperskill.org/projects/120/stages/649/implement

В пулл-реквест попал тест от Hyperskill и файлы gradle.

File: HyperMetro/task/src/metro/entity/NamedEntity.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package metro.entity;
+
+abstract class NamedEntity {
```
Comment (AndryLat) [inline HyperMetro/task/src/metro/entity/NamedEntity.java:3]: Почему видимость дефолтная? а если я где-то захочу использовать полиморфизм для этого класса?

File: HyperMetro/task/src/metro/services/MetroLoader.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        final var reader = Files.newBufferedReader(path, StandardCharsets.UTF_8);
+        final var gson = new Gson();
+
+        final Map<String, SortedMap<Integer, String>> schema = gson.fromJson(reader,
+                new TypeToken<Map<String, SortedMap<Integer, String>>>() {
+                }.getType());
+
+        if (schema.isEmpty()) {
+            throw new InvalidPropertiesFormatException("Incorrect file");
+        }
```
Comment (AndryLat) [inline HyperMetro/task/src/metro/services/MetroLoader.java:32]: Я бы это вынес в отдельный метод, тогда публичная бизнес логика была бы более читаемой.

File: HyperMetro/task/test/Tests.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
+import org.hyperskill.hstest.dynamic.DynamicTest;
+import org.hyperskill.hstest.exception.outcomes.WrongAnswer;
+import org.hyperskill.hstest.stage.StageTest;
+import org.hyperskill.hstest.testcase.CheckResult;
+import org.hyperskill.hstest.testing.TestedProgram;
+
+public class Tests extends StageTest<String> {
+    // test from the previous stage with one line
+    @DynamicTest(order = 1)
+    CheckResult test1() {
```
Comment (AndryLat) [inline HyperMetro/task/test/Tests.java:10]: Чтобы не писать комментарии как выше, лучше использовать паттерн для имени самого метода: https://dzone.com/articles/7-popular-unit-test-naming

 Если эти тесты пришли извне, тогда ок:)

Comment (rabestro) [reply HyperMetro/task/test/Tests.java:10]: Извне :)

Comment (rabestro) [reply HyperMetro/task/test/Tests.java:10]: Извне :)

File: HyperMetro/task/src/metro/services/MetroLoader.java (hunk 2 above)
Comment (rabestro) [reply HyperMetro/task/src/metro/services/MetroLoader.java:32]: Сейчас посмотрел внимательнее — точно. Здесь у нас два действия:
1) загружаем схему из файла
2) создаём объект метро по схеме

для второго пункта отдельный метод. 
Для первого пункта нет. 
Решение:
В публичном методе топ-левел логика:
schema = loadSchema(file)
return createMetro(schema)

Подробности выносим в отдельный метод loadSchema

File: HyperMetro/task/src/metro/MetroCLI.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        this.metro = metro;
+        this.commands = commands.stream()
+                .collect(Collectors.toUnmodifiableMap(Command::name, Function.identity()));
+        this.requestParser = new RequestParser(this.commands);
+    }
+
+    @Override
+    public void run() {
+        Stream.generate(ui::readLine)
+                .takeWhile(Predicate.not(EXIT::equalsIgnoreCase))
```
Comment (rabestro) [inline HyperMetro/task/src/metro/MetroCLI.java:37]: The condition for exit can be provided as a parameter for class creation

**Pull request merged**